Friends dataset

chiehminwei opened the pull request:
**Patch description**
This patch adds a new task `friends` for generating multiparty conversation training examples from the [Friends Corpus on ConvoKit](https://convokit.cornell.edu/documentation/friends.html). 

Example Training Sample:

```
Text:
Person A: I like pizza.
Person B: Me too.
Person C: Could someone order pizza.

Label:
	I will do it
```

The teacher supports 3 command line arguments, `--character`, `--use-silence-token`, and `--use-start-token`.

The `--character` option specifies which speaker labels to train on. For example, if 'Rachel Green' is specified, then only sentences uttered by Rachel Green will generate a training sample.

The choices for `character` are the main characters from the show, or an `All` option specifying all 6 main characters, excluding supporting cast, should generate training samples. The `All` option is the default option.

```
[
  'All',
  'Rachel Green',
  'Monica Geller',
  'Phoebe Buffay',
  'Joey Tribbiani',
  'Chandler Bing',
  'Ross Geller',
]
```

The `--uses-silence-token` (default: `True`) flag indicates whether a training sample should be generated with a <SILENCE> token when it's not the chosen character's turn to speak. For example, the earlier pizza conversation would generate the following training samples:

```
# Chosen Character = Person C

Text:
Person A: I like pizza.

Label:
    __SILENT__

===
Text:
Person A: I like pizza.
Person B: Me too.

Label:
      Could someone order pizza.
```

The `--uses-start-token` (default: `False`) flag indicates whether a <START> token should be included in the beginning of a conversation. For example, the earlier pizza conversation would generate the following training samples:

```
# --use_start_token = True
# Chosen Character = Person A

Text:
__Start__

Label:
  I like pizza.
```
Whereas when `--use-start-token = False`, the first sentence in a conversation would be skipped and not generate a training sample.

There are also two flags `--start-token` and `--silence-token` that you can use to specify what special symbols you want to use to represent these tokens. By default, they are `__START__` and `__SILENCE__`.


**Testing steps**

```
pytest parlai/tasks/myteacher/test.py
```

Expected Output:

```
============================================================= test session starts ==============================================================
test.py ...
```


**Other information**

To see the examples generated, run:

```
parlai display_data --task friends
```

chiehminwei commented:
@mojtaba-komeili

mojtaba-komeili commented on parlai/tasks/friends/agents.py:
```diff
@@ -0,0 +1,133 @@
+import json
+
+
+def _path(opt, *additions):
```
Do you use the additional args? If not why keeping them?

chiehminwei replied:
Ahh good catch I'll remove them

mojtaba-komeili commented on parlai/tasks/friends/agents.py:
```diff
@@ -0,0 +1,133 @@
+
+
+class DefaultTeacher(DialogTeacher):
+    START_TOKEN = '<START>'
```
For these constants, maybe it makes sense to set them via flags, that way they can be consistent when we are mixing them with other datasets that have similar tokens, See [this](https://github.com/facebookresearch/ParlAI/blob/7660197631c06ee41c0ce25c5de37a74b11ecb77/projects/blenderbot2/agents/blenderbot2.py#L279-L284).

Also, in most of our recent projects we prefer a different style for the special tokens. See the silence token [here](https://github.com/fairinternal/ParlAI-Internal/blob/1ed6496e549944cf02017930a63baa7555f9055a/projects/blenderbot3/tasks.py#L121).

mojtaba-komeili commented on parlai/tasks/friends/agents.py:
```diff
@@ -0,0 +1,133 @@
+                speaker = utterance['speaker']
+                conversation_id = utterance['conversation_id']
+
+                if conversation_id not in conversations:
```
Checkout python's defaultdict for simpler implantation here.

mojtaba-komeili reviewed (changes requested):
Thanks Jimmy, it looks good. Just a few comments to make it better:
1. try our [autoformat.sh](https://github.com/facebookresearch/ParlAI/blob/main/autoformat.sh) script for fixing the lint errors.
2. Make sure tests are passing. There are some failing now.
3. Add convokit to the installation requirements (maybe that is what causing test failures)
4. It's alright to have a single teacher, but maybe add a couple more for convenience use. They may be simple extensions of the current one with minimal changes. For example a teacher for All and another one that is a single charter POV.

mojtaba-komeili commented:
One more thing, could you add the list of existing character as another entry to the message? It could be a comma separated list of names.

chiehminwei commented:
Sounds great, all done! See updated PR. 

Used `autoformat.sh`, got around ConvoKit, added multiple convenience Teacher classes and added corresponding tests, and added cli option to specify the list of characters. All tests passed except the cleaninstall one unrelated to this PR.

mojtaba-komeili commented on parlai/core/build_data.py:
```diff
@@ -380,6 +380,8 @@ def _unzip(path, fname, delete=True):
                 PathManager.mkdirs(outpath)
                 continue
             logging.debug(f"Extracting to {outpath}")
+            if '__MACOSX' in member:
```
It logically makes sense to submit this as its own patch. Please remove it from this PR and add is as a separate one.

mojtaba-komeili commented on parlai/tasks/friends/build.py:
```diff
@@ -0,0 +1,96 @@
+
+def build(opt):
+    dpath = os.path.join(opt['datapath'], 'Friends')
+    version = '1.02'
```
You can change this to 1 since it was never published to the public code. But remove your local files to force a rebuild.

mojtaba-komeili commented:
One more thing that forgot asking you to add is sample screenshots that shows the output when you run `parlai display_data -t <your task name>`. Just a couple of teachers would be enough.

mojtaba-komeili commented:
> > 
> 
> This this what you mean? https://github.com/facebookresearch/ParlAI/pull/4568/files#r891537386

(For the record keeping) we chatted about this in-person.

chiehminwei commented:
## Screenshots

`parlai display_data --task friends`
![image](https://user-images.githubusercontent.com/2521639/172949089-6dda3b29-6bd0-453e-ba50-92b027dfdc70.png)

`parlai display_data --task friends:rachel`
![image](https://user-images.githubusercontent.com/2521639/172949158-c3f18b9f-3d02-4db2-83f3-e20af916fbc8.png)

parlai display_data --task friends:rachel --use-silence-token=False
![image](https://user-images.githubusercontent.com/2521639/172949307-5e50fbf0-51ce-4021-952e-0cc6fc381739.png)

`parlai display_data --task friends:monica`
![image](https://user-images.githubusercontent.com/2521639/172949383-a9b6e047-66d0-4342-9ad4-9777187b1dc4.png)

mojtaba-komeili reviewed (approved):
Looks good thanks. Feel free to merge.
